### ISSUE TITLE
How do I request rust-analyzer to add imports to function, not to module's top?

### ISSUE THREAD

**Issue by vi** [rust-analyzer/rust-analyzer#8103]
Typically `rust-analyzer`'s VSCode plugin inserts `use` statements at the top of the file when assisting auto-complete.

How do I make it to instead insert those `use` statements in into smaller scope?

P.S. Is Github Issues the right place to ask questions about usage of rust-analyzer?

**Comment by lnicola** [comment]
That's not possible currently possible. It could be implemented, but I'm not sure how it would work. Should it add the import to the innermost scope? To the top of the function?

**Comment by vi** [comment]
> Should it add the import to the innermost scope? To the top of the function?

Specific scope can be a configuration option. Typical values can be:

* Module (default)
* Function
* Innermost block

Ideally (ultimately) there could be some smart mode, in which imports from specific globally whitelisted crates (e.g. `std`,`core`,`alloc`,`slab`,`rand`,`libc`,`serde`,`log`) go do module scope, other imports (including from current crate) go to function scope.

Also fix suggestion for errors caused by missing imports may have multiple choices based on target scope for the use statement.

**Comment by flodiebold** [comment]
In principle we could have another, third, assist to import in the local scope; and then add assists to allow moving imports up the scopes. Another alternative would be to have an assist, triggerable on any use of a name that is imported at the top, that moves that import to the local scope :thinking:

**Comment by elijaharita** [comment]
We could leave the existing "import" code action, and add a new action "import at scope", which would function the same, but prompt the user with a list of relevant scopes as the final step (probably smallest scope first). 

This would be pretty discoverable and straightforward for users, and simple to implement. No configs required.

**Comment by DanielJoyce** [comment]
Welp RA for me now is always importing in function scope and never back in crate scope. So somewhere there is now a satting and I want to turn it off.

**Comment by DanielJoyce** [comment]
Oh, I think it's an interaction with a feature gate.

**Comment by Veykril** [comment]
Yep the current behaviour is to do function local imports if the function(same for const and statics) is cfg-attributed
